### PR TITLE
OSOE-93: Add a UI test to Lombiq.Privacy

### LINKED ISSUE #12: Add a UI test (OSOE-93)
Similar to https://github.com/Lombiq/Orchard-UIKit/issues/8 and the existing Data Tables test, add a UI test that checks the behavior of the module's features.

- There's already a test in `Lombiq.OSOCE.Tests.UI`, `ConsentBannerShouldWork`, that is a good start.
- See https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/#adding-a-new-extension-or-significant-new-features

### PR DESCRIPTION
**Pull request by dministro**
Fixes GH-12

OSOE-93

### CODE REVIEW

File: License.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Copyright © 2021, [Lombiq Technologies Ltd.](https://lombiq.com)
+Copyright © 2022, [Lombiq Technologies Ltd.](https://lombiq.com)
```
Comment (sarahelsaig) [inline License.md]: The copyright year is the "funding year" of the whole repository, which is 2018.

File: Lombiq.Privacy.Samples/Lombiq.Privacy.Samples.csproj [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+  </ItemGroup>
+
+  <PropertyGroup>
+    <Title>Lombiq Privacy for Orchard Core - Samples</Title>
+    <Authors>Lombiq Technologies</Authors>
+    <Copyright>Copyright © 2022, Lombiq Technologies Ltd.</Copyright>
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Lombiq.Privacy.Samples.csproj]: Same

File: Lombiq.Privacy.Samples/Manifest.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+[assembly: Module(
+    Name = "Lombiq Privacy - Samples",
+    Author = "Lombiq Technologies",
+    Website = "https://github.com/Lombiq/Orchard-Privacy",
+    Version = "3.1.0-alpha",
+    Description = "Samples for Lombiq Privacy.",
+    Category = "Privacy",
+    Dependencies = new[]
+    {
+        "Lombiq.HelpfulExtensions.ContentTypes",
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Manifest.cs]: Why is this hard coded? If your project depends on it, then it should be referenced, otherwise it won't become a dependency in the nuget package.

File: Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+                var workflowType = workflowTypeStore.GetAsync(workflowTypeId).GetAwaiter().GetResult();
+                var token = securityTokenService.CreateToken(
+                    new WorkflowPayload(workflowType.WorkflowTypeId, activityId),
+                    TimeSpan.FromDays(
+                        tokenLifeSpan == 0 ? HttpWorkflowController.NoExpiryTokenLifespan : tokenLifeSpan));
+
+                // LinkGenerator.GetPathByAction(...) and UrlHelper.Action(...) not resolves url for
+                // HttpWorkflowController.Invoke action.
+                // Any ideas are welcome.
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs]: This is basically a TODO comment, just without the magic word. What did you try exactly? If we can't find a solution now, please open an issue in OC and include its link here.

File: Lombiq.Privacy.Tests.UI/Constants/ElementSelectors.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+namespace Lombiq.Privacy.Tests.UI.Constants;
+
+public static class ElementSelectors
+{
+    public const string PrivacyConsentAcceptButtonId = "privacy-consent-accept-button";
+    public const string PrivacyConsentCheckboxLabel = "div.form-check > label";
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Constants/ElementSelectors.cs]: It should be `PrivacyConsentCheckboxLabelCss` because this variable contains the CSS query string for the label, not the label.

File: Lombiq.Privacy.Tests.UI/Constants/ExpectedContents.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+namespace Lombiq.Privacy.Tests.UI.Constants;
+
+public static class ExpectedContents
+{
+    public const string ConsentBannerContent = "Our site uses browser cookies to personalize your experience. "
+        + "Regarding this issue, please find our information on data control and processing here and our information on "
+        + "server logging and usage of so-called 'cookies' here. "
+        + "By clicking here You confirm your approval of the data processing activities mentioned in these documents. "
+        + "Please note that You may only use this website efficiently if you agree to these conditions.";
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Constants/ExpectedContents.cs]: Rewrap this.

File: Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        // Consent checkbox left unchecked, so after clicking on submit navigation should not happens.
+        var currentUri = context.GetCurrentUri();
+        currentUri.AbsolutePath
+            .ShouldBe(AbsolutePaths.CompetitorRegistration);
+
+        // Now we set consent checkbox to checked
+        await context.SetCheckboxValueAsync(By.Id("PrivacyConsentCheckboxPart_ConsentCheckbox"), isChecked: true);
+        await context.ClickReliablyOnSubmitAsync();
+
+        // After submit, it navigates to the new content items view
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs]: Add period to the two comments.

File: Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        // Here we fill up the competitor form.
+        await context.FillInWithRetriesAsync(By.Id("full-name"), TestCompetitor.FullName);
+        await context.FillInWithRetriesAsync(By.Id("age"), TestCompetitor.Age);
+        await context.FillInWithRetriesAsync(By.Id("e-mail"), TestCompetitor.Email);
+        await context.ClickReliablyOnSubmitAsync();
+
+        // Consent checkbox left unchecked, so after clicking on submit navigation should not happens.
+        var currentUri = context.GetCurrentUri();
+        currentUri.AbsolutePath
+            .ShouldBe(AbsolutePaths.CompetitorRegistration);
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs]: You don't really need a variable here.
```suggestion
        context.GetCurrentUri().ShouldBe(AbsolutePaths.CompetitorRegistration);
```

File: Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        // Now we set consent checkbox to checked
+        await context.SetCheckboxValueAsync(By.Id("PrivacyConsentCheckboxPart_ConsentCheckbox"), isChecked: true);
+        await context.ClickReliablyOnSubmitAsync();
+
+        // After submit, it navigates to the new content items view
+        currentUri = context.GetCurrentUri();
+        currentUri.AbsolutePath
+            .ShouldNotBe(AbsolutePaths.ErrorPage);
+        currentUri.AbsolutePath
+            .ShouldNotBe(AbsolutePaths.CompetitorRegistration);
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs]: You don't need the variable here either, if you use the `.ShouldNotBeOneOf()` extension method. It communicates the intention more clearly than individual `.ShouldNotBe()` calls.

File: Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        await context.FillInWithRetriesAsync(By.Id("Password"), TestUser.Password);
+        await context.FillInWithRetriesAsync(By.Id("ConfirmPassword"), TestUser.Password);
+        await context.SetCheckboxValueAsync(By.Id("RegistrationCheckbox"), isChecked: true);
+        await context.ClickReliablyOnSubmitAsync();
+
+        // Login with the created user
+        await context.GoToRelativeUrlAsync("/Login");
+        await context.FillInWithRetriesAsync(By.Id("UserName"), TestUser.Name);
+        await context.FillInWithRetriesAsync(By.Id("Password"), TestUser.Password);
+        await context.ClickReliablyOnSubmitAsync();
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs]: Is logging in like this necessary for the test? In my past experience it's not that reliable, so if not required then sign in directly instead.

File: Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        (await context.GetCurrentUserNameAsync()).ShouldBe(TestUser.Name);
+
+        // Check that, the consent banner not comes up after login
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs]:
```suggestion
        // Check that, the consent banner doesn't come up after login.
```
Please add period to the other comments in this file as well.

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

File: Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs (hunk 4 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs]: You can find two diffs below. One for `UrlHelper `and one for `LinkGenerator`.
The result is the same in both cases, the `url == null` at the end.

[WorkflowHelperMethodProvider_LinkGenerator.cs.diff.txt](https://github.com/Lombiq/Orchard-Privacy/files/8766487/WorkflowHelperMethodProvider_LinkGenerator.cs.diff.txt)
[WorkflowHelperMethodProvider_UrlHelper.cs.diff.txt](https://github.com/Lombiq/Orchard-Privacy/files/8766488/WorkflowHelperMethodProvider_UrlHelper.cs.diff.txt)

Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs]: Yeah, that should've worked. Please open an [OC issue](https://github.com/OrchardCMS/OrchardCore/issues) and paste the link instead of the "Any ideas are welcome." text.

Comment (dministro) [reply Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs]: I tried both cases with root ServiceProvider too.

I started [here](https://github.com/OrchardCMS/OrchardCore/blob/62eed89890b6b9b6d71a02e1465269b9bb512207/src/OrchardCore.Modules/OrchardCore.Workflows/Http/Controllers/HttpWorkflowController.cs#L71).
The `Controller.Url` is a `IUrlHelper` instance.

**Review by sarahelsaig — APPROVED** [review]

File: Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+using OrchardCore.Scripting;
+using OrchardCore.Workflows.Http.Controllers;
+using OrchardCore.Workflows.Http.Models;
+using OrchardCore.Workflows.Services;
+using System;
+using System.Collections.Generic;
+
+namespace Lombiq.Privacy.Samples.Scripting;
+
+public class WorkflowHelperMethodProvider : IGlobalMethodProvider
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Scripting/WorkflowHelperMethodProvider.cs:11]: Include some basic documentation to explain what is `workflowHttpEventUrlResolve()` and why we need it.

File: Lombiq.Privacy.Samples/Readme.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+Do you want to quickly try out this project and see it in action? Check it out in our [Open-Source Orchard Core Extensions](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions) full Orchard Core solution and also see our other useful Orchard Core-related open-source projects!
+
+## Training sections
+
+Go to Recipes in the admin dashboard and select "Lombiq Privacy - Sample Content - Privacy, Forms, Workflows".
+
+ - [Consent Banner Feature](../Readme.md#consent-banner-feature): Right after the recipe has imported, the consent banner comes up, and you can agree with it.
+ - [Registration Consent Feature](../Readme.md#registration-consent-feature): Log out first, then navigate to user registration (/Register), and you can see, the registration consent checkbox at the bottom of registration form.
+ - [Form Consent Feature](../Readme.md#form-consent-feature): Navigate to home or to /competitor-registration route, fill up the form, and you can't send the form until you don't accept the privacy policy.
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Readme.md:19]: I appreciate that there isn't much code, but at least link to the relevant parts in the recipe from this form consent feature.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 14]
```diff
@@ -0,0 +1,504 @@
...
+            ]
+          }
+        }
+      ]
+
+    },
+    {
+      "name": "Settings",
+      "RegistrationSettings": {
+        "UsersCanRegister": 1,
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Why is this `1` instead of `true`?

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 15]
```diff
@@ -0,0 +1,504 @@
...
+        "UsersMustValidateEmail": false,
+        "UsersAreModerated": false,
+        "UseSiteTheme": false,
+        "NoPasswordForExternalUsers": false,
+        "NoUsernameForExternalUsers": false,
+        "NoEmailForExternalUsers": false,
+        "UseScriptToGenerateUsername": false,
+        "GenerateUsernameScript": "/* Uncomment to map AzureAD\r\n// Uncomment to output the context variable in the logs\r\n// log(\"warning\", JSON.stringify(context));\r\nswitch (context.loginProvider) {\r\n    case \"AzureAD\":\r\n        context.userName = \"azad\" + uuid();\r\n        break;\r\n    default:\r\n        log(\"Warning\", \"Provider \" + context.loginProvider + \" was not handled\");\r\n        // Uncomment to generate a username as a uuid\r\n        // context.userName = \"u\" + uuid();\r\n        break;\r\n}\r\n*/\r\n"
+      }
+    },
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Why is this necessary, and is specifying all of it necessary? Please add a comment above it explaining the reason.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 16]
```diff
@@ -0,0 +1,504 @@
...
+          ]
+        }
+      ]
+    },
+    {
+      "name": "content",
+      "data": [
+        {
+          "ContentItemId": "4ftg1gchvpstvt5zz4hhnr84w0",
+          "ContentItemVersionId": "491rv1zfbdvbxxzcbvgbks4mn4",
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Please give these both Page content items human readable content item ID, and `[js:uuid()]` version ID.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 17]
```diff
@@ -0,0 +1,504 @@
+{
+  "name": "Lombiq.Privacy.Samples.Content",
+  "displayName": "Lombiq Privacy - Sample Content - Privacy, Forms, Workflows",
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: I think it would be more manageable to split into 3 files:
- features and settings
- content and workflow definitions
- pages

Right now it's quite daunting.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 18]
```diff
@@ -0,0 +1,504 @@
...
+    {
+      "name": "Settings",
+      "HomeRoute": {
+        "Action": "Display",
+        "Controller": "Item",
+        "Area": "OrchardCore.Contents",
+        "contentItemId": "4fxbwpz7p0gqntyxxd31zmqk4p",
+        "jsonPath": ""
+      }
+    }
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Is it a good idea to alter such a core setting? Now after running the recipe the site behaves more differently then you'd expect.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 19]
```diff
@@ -0,0 +1,504 @@
...
+            "Disabled": false,
+            "RouteContainedItems": false,
+            "Absolute": false
+          },
+          "FlowPart": {
+            "Widgets": []
+          }
+        },
+        {
+          "ContentItemId": "4fxbwpz7p0gqntyxxd31zmqk4p",
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Create a `MainMenuNavigationProviderBase` implementation with a link that points to this page and perhaps a link to the workflow on the admin dashboard as well.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json [hunk 20]
```diff
@@ -0,0 +1,504 @@
...
+              "SourceActivityId": "474fg602kmn9z0xsn389w51mnf",
+              "SourceOutcomeName": "Failed",
+              "DestinationActivityId": "4n7zznh3ssm9078wyva0rq45d2"
+            }
+          ]
+        }
+      ]
+    },
+    {
+      "name": "content",
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]:
```suggestion
      "name": "Content",
```

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 14 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: `UserCanRegister` is an enum and not a boolean property.
See in [RegistrationSettings.cs](https://github.com/OrchardCMS/OrchardCore/blob/62eed89890b6b9b6d71a02e1465269b9bb512207/src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs#L5)

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 15 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: This is a sample project to represent the capabilities of `Lombiq.Privacy` module.
I think, setting all the site, etc. properties is the best solution to reproduce a clean environment. Who knows what happened with the site before.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 18 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: This is the main "function" of the sample.
To represent the usage of `Form consent` feature, I created a possible solution for a use case, where competitors can make registration, and the registration form comes up as home page.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 20 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: This is lower case in documentation. See [here](https://docs.orchardcore.net/en/dev/docs/reference/modules/Recipes/#content-step)

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 14 above)
Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: I see, that makes sense. However you can set an enum with string too which is much less cryptic. I have just verified that this works:
```suggestion
        "UsersCanRegister": "AllowRegistration",
```

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 15 above)
Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Okay, but then include a comment explaining this. If I understand correctly all of these are the default settings, except "UsersCanRegister", right? Make sure to indicate which values are customized.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 14 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: You are right.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 18 above)
Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: I don't really understand this. It's just the home page, you can still access other pages without interacting with this form. So why can't you access this page from the menu or via direct link instead of the home page?

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 20 above)
Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: If you look around on the same page ["feature"](https://docs.orchardcore.net/en/dev/docs/reference/modules/Recipes/#feature-step) is lower case too, yet you used "Feature" in the recipe. I'm pretty sure it's case-insensitive. (I have just verified that using "Content" works.) We should make it consistent and all the other step names are PascalCase so this should be as well.

File: Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json (hunk 17 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: I think, I'm facing an OC issue again.

When I split the recipe into multiple files, it throws a `JavaScriptException: variables is not defined` if a `[js: variables('errorContentItemId')]` expression is present in the `InnerRecipe`...

Example to reproduce:

`recipe 1`:
```json
{
  "name": "Lombiq.Privacy.Samples",
  "displayName": "Lombiq Privacy - Sample - Privacy, Forms, Workflows",
  "description": "Sample for testing and demonstration of the Lombiq Privacy module.",
  "author": "Lombiq Technologies",
  "website": "https://github.com/Lombiq/Orchard-Privacy",
  "version": "1.0",
  "issetuprecipe": false,
  "categories": [
    "training"
  ],
  "tags": [
    "developer",
    "training",
    "sample"
  ],
  "variables": {
    "errorContentItemId": "erroritem00000000000000000",
    "registrationContentItemId": "registrationitem0000000000",
    "registrationWorkflowTypeId": "registrationworkflow000000",
    "registrationHttpRequestActivityId": "registrationhttpevent00000"
  },
  "steps": [
    {
      "name": "Recipes",
      "Values": [
        {
          "executionid": "Lombiq.Privacy.Samples",
          "name": "Lombiq.Privacy.Samples.Content"
        }
      ]
    }
  ]
}
```

`recipe 2`:
```json
{
  "name": "Lombiq.Privacy.Samples.Content",
  "displayName": "Lombiq Privacy - Sample - Privacy, Forms, Workflows",
  "description": "Sample for testing and demonstration of the Lombiq Privacy module.",
  "author": "Lombiq Technologies",
  "website": "https://github.com/Lombiq/Orchard-Privacy",
  "version": "1.0",
  "issetuprecipe": false,
  "categories": [
    "training"
  ],
  "tags": [
    "hidden"
  ],
  "steps": [
    {
      "name": "Content",
      "data": [
        {
          "ContentItemId": "[js: variables('errorContentItemId')]",
          "ContentItemVersionId": "[js: uuid()]",
          ...
        },
        ...
      ]
    }
  ]
}
```

If I copy the `Content` step from `recipe 2` into `recipe 1` everything works fine.

Did you faced something similar before?

Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: Yep, the `variables: {...}` object should be defined in `InnerRecipe` too, so they are not global

Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: That's unfortunate. Then keep it as one recipe.

Comment (dministro) [reply Lombiq.Privacy.Samples/Recipes/Lombiq.Privacy.Samples.Content.recipe.json]: No, I can split the recipe into multiple files, the exception message was a bit confusing, the problem was not the script method `variables(...)` but the scope/visibility of the `"variables": {...}` object in the main recipe.
If I define the `"variables": {...}` inside `InnerRecipe` it works as expected.

File: Lombiq.Privacy.Samples/Navigation/PrivacySamplesNavigationProvider.cs [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+        IHttpContextAccessor hca,
+        IStringLocalizer<PrivacySamplesNavigationProvider> stringLocalizer)
+        : base(hca, stringLocalizer)
+    {
+    }
+
+    protected override void Build(NavigationBuilder builder) =>
+        builder
+            .Add(T["Privacy Samples"], builder => builder
+                .Add(T["Competitor registration"], itemBuilder => itemBuilder.Url("/competitor-registration")));
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Navigation/PrivacySamplesNavigationProvider.cs]: I meant put both links into here, so all the relevant pages are available from the same main menu hub. You don't need the separate admin menu.

File: Lombiq.Privacy.Samples/Lombiq.Privacy.Samples.csproj [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="OrchardCore.Forms" Version="1.3.0" />
+    <PackageReference Include="OrchardCore.Module.Targets" Version="1.3.0" />
+    <PackageReference Include="OrchardCore.Workflows" Version="1.3.0" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\Lombiq.HelpfulExtensions\Lombiq.HelpfulExtensions.csproj" />
```
Comment (sarahelsaig) [inline Lombiq.Privacy.Samples/Lombiq.Privacy.Samples.csproj:35]: This should be  `Condition="'$(NuGetBuild)' != 'true'"` and have a nuget package counterpart.

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

File: Lombiq.Privacy.Samples/Navigation/PrivacySamplesNavigationProvider.cs (hunk 21 above)
Comment (dministro) [reply Lombiq.Privacy.Samples/Navigation/PrivacySamplesNavigationProvider.cs]: I don't understand. Why we want to bring an admin function out to a non admin menu?

Comment (sarahelsaig) [reply Lombiq.Privacy.Samples/Navigation/PrivacySamplesNavigationProvider.cs]: This is a demo site, the purpose of the menu is to guide a user to the points of interest for that module so they don't have to hunt through the source code to find what they need to look for.

File: Lombiq.Privacy/Manifest.cs [hunk 23]
```diff
@@ -5,7 +5,7 @@
     Name = "Lombiq Privacy",
     Author = "Lombiq Technologies",
     Website = "https://lombiq.com",
-    Version = "3.0.1"
+    Version = "3.1.0-alpha"
```
Comment (Piedone) [inline Lombiq.Privacy/Manifest.cs]: Next time don't skip a minor version when publishing an alpha (the current version at the time was 3.0.1 so this should have been 3.0.2-alpha): https://lombiq.atlassian.net/wiki/spaces/DEV/pages/786857987/Managing+releases+of+open-source+projects#:~:text=When%20publishing%20a%20pre%2Drelease%20version%20for%20something%20temporary%20you%20do%20under%20an%20issue%2C%20then%20always%20use%20the%20next%20patch%20version%20with%20a%20suffix.%20E.g.%20if%20the%20current%20latest%20released%20version%20if%201.2.3%20then%20publish%201.2.4%2Dalpha.0.
